go/adbc/driver/flightsql/record_reader.go

```diff
@@ -32,38 +32,48 @@ import (
 )
 
 type reader struct {
-	refCount int64
-	schema   *arrow.Schema
-	ch       chan arrow.Record
-	rec      arrow.Record
-	err      error
+	refCount   int64
+	schema     *arrow.Schema
+	chs        []chan arrow.Record
+	curChIndex int
+	rec        arrow.Record
+	err        error
 
 	cancelFn context.CancelFunc
 }
 
 // kicks off a goroutine for each endpoint and returns a reader which
 // gathers all of the records as they come in.
 func newRecordReader(ctx context.Context, alloc memory.Allocator, cl *flightsql.Client, info *flight.FlightInfo, clCache gcache.Cache) (rdr array.RecordReader, err error) {
-	ch := make(chan arrow.Record, 5)
-	group, ctx := errgroup.WithContext(ctx)
-	ctx, cancelFn := context.WithCancel(ctx)
-
-	defer func() {
+	endpoints := info.Endpoint
+	numEndpoints := len(endpoints)
+	var schema *arrow.Schema
+	if len(endpoints) == 0 {
+		if info.Schema == nil {
+			return nil, adbc.Error{
+				Msg:  "Server returned FlightInfo with no schema and no endpoints, cannot read stream",
+				Code: adbc.StatusInternal,
+			}
+		}
+		schema, err = flight.DeserializeSchema(info.Schema, alloc)
 		if err != nil {
-			cancelFn()
-			for rec := range ch {
-				rec.Release()
+			return nil, adbc.Error{
+				Msg:  "Server returned FlightInfo with invalid schema and no endpoints, cannot read stream",
+				Code: adbc.StatusInternal,
 			}
 		}
-	}()
+		return array.NewRecordReader(schema, []arrow.Record{})
+	}
 
-	endpoints := info.Endpoint
+	ch := make(chan arrow.Record, 5)
+	group, ctx := errgroup.WithContext(ctx)
+	ctx, cancelFn := context.WithCancel(ctx)
 
-	var schema *arrow.Schema
 	if info.Schema != nil {
 		schema, err = flight.DeserializeSchema(info.Schema, alloc)
 		if err != nil {
 			close(ch)
+			cancelFn()
```

Comment:
instead of making sure we do this everywhere we return an error, you could use a defer

```go
defer func() {
    if err != nil {
        close(ch)
        cancelFn()
    }
}()
```

that way we don't have to worry about forgetting a branch.